internal/api/iterator.go

```diff
@@ -53,39 +53,74 @@ func endCall(callID uint64) {
 	}
 }
 
-// storeIterator will add this to the end of the frame for the given ID and return a reference to it.
-// We start counting with 1, so the 0 value is flagged as an error. This means we must
-// remember to do idx-1 when retrieving
+// storeIterator will add this to the end of the frame for the given call ID and return
+// an interator ID to reference it.
+//
+// We assign iterator IDs starting with 1 for historic reasons. This could be changed to 0
+// I guess.
 func storeIterator(callID uint64, it types.Iterator, frameLenLimit int) (uint64, error) {
 	iteratorFramesMutex.Lock()
 	defer iteratorFramesMutex.Unlock()
 
-	old_frame_len := len(iteratorFrames[callID])
-	if old_frame_len >= frameLenLimit {
+	new_index := len(iteratorFrames[callID])
+	if new_index >= frameLenLimit {
 		return 0, fmt.Errorf("Reached iterator limit (%d)", frameLenLimit)
 	}
 
-	// store at array position `old_frame_len`
+	// store at array position `new_index`
 	iteratorFrames[callID] = append(iteratorFrames[callID], it)
-	new_index := old_frame_len + 1
 
-	return uint64(new_index), nil
+	iterator_id, ok := indexToIteratorID(new_index)
+	if !ok {
+		// This error case is not expected to happen since the above code ensures the
+		// index is in the range [0, frameLenLimit-1]
+		return 0, fmt.Errorf("Could not convert index to iterator ID")
+	}
+	return iterator_id, nil
 }
 
-// retrieveIterator will recover an iterator based on index. This ensures it will not be garbage collected.
-// We start counting with 1, in storeIterator so the 0 value is flagged as an error. This means we must
-// remember to do idx-1 when retrieving
-func retrieveIterator(callID uint64, index uint64) types.Iterator {
+// retrieveIterator will recover an iterator based on its ID.
+func retrieveIterator(callID uint64, iteratorID uint64) types.Iterator {
+	indexInFrame, ok := iteratorIdToIndex(iteratorID)
+	if !ok {
+		return nil
+	}
+
 	iteratorFramesMutex.Lock()
 	defer iteratorFramesMutex.Unlock()
 	myFrame := iteratorFrames[callID]
 	if myFrame == nil {
 		return nil
 	}
-	posInFrame := int(index) - 1
-	if posInFrame < 0 || posInFrame >= len(myFrame) {
+	if indexInFrame >= len(myFrame) {
 		// index out of range
 		return nil
 	}
-	return myFrame[posInFrame]
+	return myFrame[indexInFrame]
+}
+
+const (
+	INT32_MAX_AS_UINT64 uint64 = 2147483647
+	INT32_MAX_AS_INT    int    = 2147483647
+)
```

Comment:
No need for these. You can just use `math.MaxInt32`. It also works with different integer types, so no need for separate constants for each type.

Reply:
type "untyped int" – wow, not bad 🏅